pkg/kubernetes/integration_test.go

```diff
@@ -0,0 +1,308 @@
+		_, err = k8s.Client().CoreV1().Pods(namespace).Create(
+			context.TODO(),
+			&nginx,
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			t.Fatalf("failed to create pod: %v", err)
+		}
+
+		// wait for the pod to be ready for accepting requests
+		running, err := k8s.PodHelper(namespace).WaitPodRunning(context.TODO(), "nginx", time.Second*20)
+		if err != nil {
+			t.Fatalf("error waiting for pod %v", err)
+		}
+		if !running {
+			t.Fatalf("timeout expired waiting for pod ready")
+		}
+	})
+
+	// Test Wait Service Ready helper
+	t.Run("Wait Service Ready", func(t *testing.T) {
+		t.Parallel()
+
+		namespace, err := createRandomTestNamespace(k8s)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(func() {
+			_ = k8s.Client().CoreV1().Namespaces().Delete(context.TODO(), namespace, metav1.DeleteOptions{})
+		})
+
+		// Deploy nginx and expose it as a service. Intentionally not using e2e fixures
+		// because these functions rely on WaitPodRunning and WaitServiceReady which we
+		// are testing here.
+		nginxPod := fixtures.BuildNginxPod()
+		_, err = k8s.Client().CoreV1().Pods(namespace).Create(
+			context.TODO(),
+			&nginxPod,
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			t.Fatalf("failed to create pod: %v", err)
+		}
+
+		nginxSvc := fixtures.BuildNginxService()
+		_, err = k8s.Client().CoreV1().Services(namespace).Create(
+			context.TODO(),
+			&nginxSvc,
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			t.Fatalf("failed to create nginx service: %v", err)
+		}
+
+		// wait for the service to be ready for accepting requests
+		err = k8s.ServiceHelper(namespace).WaitServiceReady(context.TODO(), "nginx", time.Second*20)
+		if err != nil {
+			t.Fatalf("error waiting for service nginx: %v", err)
+		}
+	})
+
+	t.Run("Exec Command", func(t *testing.T) {
+		t.Parallel()
+
+		namespace, err := createRandomTestNamespace(k8s)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(func() {
+			_ = k8s.Client().CoreV1().Namespaces().Delete(context.TODO(), namespace, metav1.DeleteOptions{})
+		})
+
+		busybox := fixtures.BuildBusyBoxPod()
+		_, err = k8s.Client().CoreV1().Pods(namespace).Create(
+			context.TODO(),
+			&busybox,
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			t.Fatalf("failed to create pod: %v", err)
+		}
+
+		// wait for the pod to be ready
+		running, err := k8s.PodHelper(namespace).WaitPodRunning(context.TODO(), "busybox", time.Second*20)
+		if err != nil {
+			t.Fatalf("error waiting for pod %v", err)
+		}
+		if !running {
+			t.Fatalf("timeout expired waiting for pod ready")
+		}
+
+		greetings := "hello world"
+		cmd := []string{"echo", "-n", greetings}
+		stdout, _, err := k8s.PodHelper(namespace).Exec(
+			context.TODO(),
+			"busybox",
+			"busybox",
+			cmd,
+			nil,
+		)
+		if err != nil {
+			t.Fatalf("error executing command in pod: %v", err)
+		}
+
+		if string(stdout) != greetings {
+			t.Fatalf("stdout does not match expected result:\nexpected: %s\nactual%s\n", greetings, string(stdout))
```

Comment:
I think `%q` could be a bit more helpful if there were a difference with newlines or whitespaces.

```suggestion
			t.Fatalf("stdout does not match expected result:\nexpected: %q\nactual%q\n", greetings, string(stdout))
```